### PR TITLE
Less memory intensive Keeper startup

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Keeper improvement: improve memory-usage during startup by delaying log preprocessing.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 2d851ec16316f4e03c0439f685b34fb026d609a5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55660/2d851ec16316f4e03c0439f685b34fb026d609a5/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6530940858">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55660/2d851ec16316f4e03c0439f685b34fb026d609a5/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by antonio2368** [comment]
> test_postgresql_replica_database_engine_2/test.py::test_replica_consumer

@kssenii based on a quick check seems like it started failing after this PR https://github.com/ClickHouse/ClickHouse/pull/55145?

File: src/Coordination/KeeperDispatcher.cpp [hunk 1]
```diff
@@ -445,11 +452,17 @@ void KeeperDispatcher::shutdown()
...
+            if (keeper_context->shutdown_called.exchange(true))
                 return;
 
             LOG_DEBUG(log, "Shutting down storage dispatcher");
 
+            our_last_committed_log_idx = std::numeric_limits<uint64_t>::max();
+            our_last_committed_log_idx.notify_all();
+
+            keeper_context->local_logs_preprocessed = true;
+            keeper_context->local_logs_preprocessed.notify_all();
```
Comment (alesapin) [inline src/Coordination/KeeperDispatcher.cpp:464]: I'm slightly afraid of atomic synchronization, but do we wait somewhere for this flag value?

File: src/Coordination/KeeperContext.h [hunk 2]
```diff
@@ -54,6 +54,8 @@ class KeeperContext
 
     constexpr KeeperDispatcher * getDispatcher() const { return dispatcher; }
 
+    std::atomic<bool> local_logs_preprocessed = false;
```
Comment (alesapin) [inline src/Coordination/KeeperContext.h:57]: Need comment

File: src/Coordination/KeeperServer.cpp [hunk 3]
```diff
@@ -619,6 +613,82 @@ nuraft::cb_func::ReturnCode KeeperServer::callbackFunc(nuraft::cb_func::Type typ
...
+
+                auto last_committed_index = state_machine->last_commit_index();
+                // Actual log number.
+                auto index_to_commit = std::min({last_log_idx_on_disk, req.get_last_log_idx(), req.get_commit_idx()});
+
+                if (index_to_commit > last_committed_index)
+                {
+                    LOG_TRACE(log, "Trying to commit local log entries, committing upto {}", index_to_commit);
+                    raft_instance->commitLogs(index_to_commit, true);
+                    if (!keeper_context->local_logs_preprocessed)
```
Comment (alesapin) [inline src/Coordination/KeeperServer.cpp:670]: It fills like this if should be above previous line

**Review by alesapin — APPROVED** [review]

File: src/Coordination/KeeperDispatcher.cpp (hunk 1 above)
Comment (antonio2368) [reply src/Coordination/KeeperDispatcher.cpp:464]: needs to be removed, leftover from previous implementation

File: src/Coordination/KeeperServer.cpp (hunk 3 above)
Comment (antonio2368) [reply src/Coordination/KeeperServer.cpp:670]: We manually commit in this thread all the local logs which we can. All the other logs should be preprocessed inside the callback during the commit. This basically asserts that we don't have any local log uncommitted or unprocessed.

**Pull request merged**